component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateAuthenticator.java

```diff
@@ -325,4 +376,97 @@ private void allowUser(String userName, Map claims, X509Certificate cert,
+                        X509CertificateConstants.X509_CERTIFICATE_ALTERNATIVE_NAMES_NOTFOUND_ERROR_CODE);
+                throw new AuthenticationFailedException(
+                        X509CertificateConstants.X509_CERTIFICATE_ALTERNATIVE_NAMES_NOTFOUND_ERROR);
+            }
+        } catch (CertificateParsingException | IOException e) {
+            throw new AuthenticationFailedException("Failed to Parse the certificate");
+        }
+        if (matches.isEmpty()) {
+            authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                    X509CertificateConstants.X509_CERTIFICATE_ALTERNATIVE_NAMES_REGEX_NO_MATCHES_ERROR_CODE);
+            throw new AuthenticationFailedException("Regex Configured but no matches found for the given regex");
+        } else if (matches.size() > 1) {
+            authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                    X509CertificateConstants.X509_CERTIFICATE_ALTERNATIVE_NAMES_REGEX_MULTIPLE_MATCHES_ERROR_CODE);
+            throw new AuthenticationFailedException("More than one match for the given regex");
+        } else if (matches.size() == 1) {
+            return matches.get(0);
+        } else {
+            return null;
+        }
+    }
+
+    private void validateUsingSubject(String subject, AuthenticationContext authenticationContext,
+            X509Certificate cert, Map<ClaimMapping, String> claims) throws AuthenticationFailedException {
+
+        byte[] data;
+        try {
+            data = cert.getEncoded();
+        } catch (CertificateEncodingException e) {
+            throw new AuthenticationFailedException("Encoded certificate in not found", e);
+        }
+        AuthenticatedUser authenticatedUser = getUsername(authenticationContext);
+
+        if (log.isDebugEnabled()) {
+            log.debug("Getting X509Certificate username");
+        }
+        if (authenticatedUser != null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Authenticated username is: " + authenticatedUser);
+            }
+            String authenticatedUserName = authenticatedUser.getAuthenticatedSubjectIdentifier();
+            if (authenticatedUserName.equals(subject)) {
+                addOrValidateCertificate(subject, authenticationContext, data, claims, cert);
+            } else {
+                authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                        X509CertificateConstants.USERNAME_CONFLICT);
+                throw new AuthenticationFailedException(
+                        "Couldn't find X509 certificate to this authenticated user: " + authenticatedUserName);
+            }
+        } else {
+            addOrValidateCertificate(subject, authenticationContext, data, claims, cert);
+        }
+    }
+
+    private List getMatchedString(Matcher matcher, List matches) {
```

Comment:
Method name doesn't match with what it does. IMO it is **_addMatchStringToList_**. Use proper naming. Also, return of List is not essential. Simply add to list and make void

Reply:
done